client/asset/eth/eth.go

```diff
@@ -338,15 +337,92 @@ func (eth *ExchangeWallet) OwnsAddress(address string) (bool, error) {
 
 // Balance returns the total available funds in the account. The eth node
 // returns balances in wei. Those are flored and stored as gwei, or 1e9 wei.
-//
-// TODO: Return Immature and Locked values.
 func (eth *ExchangeWallet) Balance() (*asset.Balance, error) {
 	eth.lockedFundsMtx.Lock()
 	defer eth.lockedFundsMtx.Unlock()
 
 	return eth.balanceImpl()
 }
 
+// txEffectOnBalance calculates the effects of a transaction on the balance
+// of the sender. It only takes into account the transaction types expected
+// to be sent by this wallet, including withdrawals, initiations, redeems,
+// and refunds.
+func (eth *ExchangeWallet) txEffectOnBalance(tx *types.Transaction) (incoming *big.Int, outgoing *big.Int, err error) {
+	incoming, outgoing = new(big.Int), new(big.Int)
+
+	// Max possible gas fee
+	feeCap := tx.GasFeeCap()
+	gas := big.NewInt(int64(tx.Gas()))
+	outgoing.Mul(feeCap, gas)
+
+	// If the tx contains value, it is either an initiation or withdrawal.
+	// In either case we don't need to parse the tx data, so we can return
+	// here.
+	txValue := tx.Value()
+	if txValue.Cmp(new(big.Int)) > 0 {
+		outgoing.Add(outgoing, txValue)
+		return incoming, outgoing, nil
+	}
+
+	if redemptions, err := dexeth.ParseRedeemData(tx.Data()); err == nil {
+		for _, redemption := range redemptions {
+			swap, err := eth.node.swap(eth.ctx, eth.acct, redemption.SecretHash)
+			if err != nil {
+				return nil, nil, err
```

Comment:
Wondering if log and continue would be better here so the balance isn't wrecked.  We can just see how this goes though.